onnxruntime/core/providers/qnn/builder/opbuilder/gather_op_builder.cc

```diff
@@ -68,38 +126,39 @@ Status GatherOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
 
   if (is_initializer_input) {
     const auto& input_tensor = qnn_model_wrapper.GetInitializerTensors().at(input_name);
+    std::vector<uint8_t> unpacked_tensor;
+
     ORT_RETURN_IF_ERROR(qnn_model_wrapper.UnpackInitializerData(*input_tensor, unpacked_tensor));
+
     if (qnn_data_type == QNN_DATATYPE_INT_64) {
-      // Convert initializer from int64 to int32
-      size_t size = unpacked_tensor.size() / sizeof(int64_t);
-      const int64_t* gather_indices_int64 = reinterpret_cast<const int64_t*>(unpacked_tensor.data());
-      gather_indices.resize(size * sizeof(int32_t));
-      int32_t* gather_indices_int32 = reinterpret_cast<int32_t*>(gather_indices.data());
-      std::transform(gather_indices_int64, gather_indices_int64 + size, gather_indices_int32,
-                     [](int64_t item) { return SafeInt<uint32_t>(item); });
+      if (desired_data_type == QNN_DATATYPE_INT_32) {
+        ConvertInt64IndicesBytes<int32_t>(unpacked_tensor, gather_indices);
+      } else {
+        ConvertInt64IndicesBytes<uint32_t>(unpacked_tensor, gather_indices);
+      }
     } else {
       gather_indices = std::move(unpacked_tensor);
```

Comment:
What about qnn_data_type == QNN_DATATYPE_INT_32 && desired_data_type == QNN_DATATYPE_UINT_32?
The existing code for Gather OP always convert indices to signed int32, because Onnx always use signed int.